File: src/shipyard/ship/wait.py
"""Truth-condition evaluators for ``shipyard wait``.

Each evaluator takes a snapshot from ``gh`` and returns a deterministic
``TruthResult`` describing:

* ``matched`` — True if the wait condition is satisfied.
* ``observed`` — a JSON-serializable view of what was observed (populated
  into the ``OutputEnvelope.observed`` field).
* ``error`` — a caller-friendly sentinel when evaluation can't proceed
  (invalid PR number, rulesets detected, etc.).

Pure logic — no I/O. The CLI layer handles the actual ``gh`` invocation
(via ``wait_transport``) so these stay unit-testable against fixture
JSON.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

# Conclusion mapping mirrors docs/shipyard-wait-primitive-design-v4.md.
PASSING_CONCLUSIONS = frozenset({"SUCCESS", "NEUTRAL", "SKIPPED"})
FAILING_CONCLUSIONS = frozenset(
    {"FAILURE", "TIMED_OUT", "CANCELLED", "ACTION_REQUIRED", "STARTUP_FAILURE", "STALE"}
)
STILL_WAITING_STATES = frozenset({"QUEUED", "IN_PROGRESS", "PENDING"})


class UnsupportedScopeError(Exception):
    """Raised when a condition can't be evaluated because it sits in a
    governance surface we don't support yet (rulesets, merge queues).

    The CLI surfaces this as exit code 7.
    """


class InvalidInputError(Exception):
    """Raised when input refers to a missing PR, release, or run.

    The CLI surfaces this as exit code 5.
    """


@dataclass(frozen=True)
class TruthResult:
    matched: bool
    observed: dict[str, Any] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# release


def evaluate_release(
    snapshot: dict[str, Any] | None,
    *,
    manifest: list[str] | None,
) -> TruthResult:
    """Matches when the release exists, isn't a draft, and every asset in
    the manifest has ``state == "uploaded"``. When ``manifest`` is None
    (or empty), any non-empty assets list counts as matched.

    ``snapshot`` is the ``gh api`` response for
    ``repos/{owner}/{repo}/releases/tags/{tag}``. Pass ``None`` when the
    release doesn't exist yet — returns ``matched=False`` so callers can
    keep waiting.
    """
    if snapshot is None:
        return TruthResult(matched=False, observed={"exists": False})
    if snapshot.get("draft") is True:
        return TruthResult(
            matched=False,
            observed={"exists": True, "draft": True},
        )
    assets_raw = snapshot.get("assets") or []
    assets: list[dict[str, Any]] = []
    for asset in assets_raw:
        if isinstance(asset, dict):
            assets.append(
                {
                    "name": str(asset.get("name", "")),
                    "state": str(asset.get("state", "")),
                    "size": int(asset.get("size", 0) or 0),
                }
            )
    observed: dict[str, Any] = {
        "exists": True,
        "draft": False,
        "tag_name": snapshot.get("tag_name", ""),
        "assets": assets,
    }
    if manifest:
        by_name = {a["name"]: a for a in assets}
        missing: list[str] = []
        not_uploaded: list[str] = []
        for name in manifest:
            entry = by_name.get(name)
            if entry is None:
                missing.append(name)
            elif entry["state"] != "uploaded":
                not_uploaded.append(name)
        observed["manifest"] = list(manifest)
        observed["missing"] = missing
        observed["not_uploaded"] = not_uploaded
        matched = not missing and not not_uploaded
        return TruthResult(matched=matched, observed=observed)
    # No manifest: match as soon as any asset is uploaded.
    has_upload = any(a["state"] == "uploaded" for a in assets)
    return TruthResult(matched=has_upload, observed=observed)


# ---------------------------------------------------------------------------
# pr --state green


def evaluate_pr_green(snapshot: dict[str, Any] | None) -> TruthResult:
    """Matches when every classic-branch-protection required check on the
    PR's current head SHA has conclusion ∈ {SUCCESS, NEUTRAL, SKIPPED}.

    ``snapshot`` is the ``gh pr view --json
    number,headRefOid,state,mergeable,statusCheckRollup,mergeStateStatus``
    response. Rulesets / merge-queue flavoured PRs raise
    :class:`UnsupportedScopeError` — the caller maps that to exit code 7.
    """
    if snapshot is None:
        raise InvalidInputError("PR not found")
    merge_state = str(snapshot.get("mergeStateStatus") or "").upper()
    # GitHub's mergeStateStatus returns NOT_SUPPORTED/BLOCKED variants
    # when rulesets / merge-queue gate required checks. Detect on the
    # distinctive name-rich rulesets flag rather than a plain BLOCKED
    # (BLOCKED is also used for branch-protection-required checks that
    # simply haven't landed yet, which is our happy path).
    if "RULESET" in merge_state or merge_state == "MERGE_QUEUED":
        raise UnsupportedScopeError(
            "Rulesets / merge-queue governance isn't supported by "
            "`shipyard wait pr --state green` yet — see "
            "governance/profiles.py."
        )
    rollup = snapshot.get("statusCheckRollup") or []
    head_sha = str(snapshot.get("headRefOid") or "")
    required_entries: list[dict[str, Any]] = []
    advisory_entries: list[dict[str, Any]] = []
    all_required_pass = True
    any_still_waiting = False
    for entry in rollup:
        if not isinstance(entry, dict):
            continue
        name = str(entry.get("name") or entry.get("context") or "")
        state = str(entry.get("state") or "").upper()
        conclusion = str(entry.get("conclusion") or "").upper()
        required = bool(entry.get("isRequired", False))
        observed_entry = {
            "name": name,
            "state": state or None,
            "conclusion": conclusion or None,
            "required": required,
        }
        if not required:
            advisory_entries.append(observed_entry)
            continue
        required_entries.append(observed_entry)
        if state in STILL_WAITING_STATES and conclusion not in PASSING_CONCLUSIONS:
            any_still_waiting = True
            all_required_pass = False
            continue
        if conclusion in PASSING_CONCLUSIONS:
            continue
        # Terminal-but-not-passing → never-matching. Broadcast as
        # a non-match so waiters can exit 1 on --timeout (not 0).
        all_required_pass = False
    if not required_entries:
        # No required checks is a valid matched state only when
        # mergeable is clean; otherwise report as non-match so callers
        # keep polling.
        all_required_pass = str(snapshot.get("mergeable") or "").upper() == "MERGEABLE"
    matched = all_required_pass and not any_still_waiting
    return TruthResult(
        matched=matched,
        observed={
            "pr": snapshot.get("number"),
            "head_sha": head_sha,
            "merge_state_status": merge_state or None,
            "checks": required_entries,
            "advisory": advisory_entries,
        },
    )


# ---------------------------------------------------------------------------
# pr --state merged / closed


def evaluate_pr_state(
    snapshot: dict[str, Any] | None,
    *,
    target_state: str,
) -> TruthResult:
    """Matches when PR's state reaches ``merged`` or ``closed``."""
    if snapshot is None:
        raise InvalidInputError("PR not found")
    state = str(snapshot.get("state") or "").upper()
    merged = bool(snapshot.get("merged", False))
    observed = {
        "pr": snapshot.get("number"),
        "state": state,
        "merged": merged,
    }
    if target_state == "merged":
        return TruthResult(matched=merged, observed=observed)
    if target_state == "closed":
        return TruthResult(matched=state in {"CLOSED", "MERGED"}, observed=observed)
    raise InvalidInputError(f"unknown target state {target_state!r}")


# ---------------------------------------------------------------------------
# run


RUN_TERMINAL_STATUSES = frozenset({"completed"})


def evaluate_run(
    snapshot: dict[str, Any] | None,
    *,
    require_success: bool,
) -> TruthResult:
    """Matches when a workflow run reaches a terminal status.

    With ``--success`` the match additionally requires ``conclusion ==
    "success"``. Any other terminal conclusion is a fail-fast: we raise
    a :class:`RunFailedFastError` so the CLI can exit 4 rather than wait out
    the timeout on a run that will never pass.
    """
    if snapshot is None:
        raise InvalidInputError("run not found")
    status = str(snapshot.get("status") or "").lower()
    conclusion = str(snapshot.get("conclusion") or "").lower()
    observed = {
        "run_id": snapshot.get("databaseId") or snapshot.get("id"),
        "status": status,
        "conclusion": conclusion or None,
    }
    if status not in RUN_TERMINAL_STATUSES:
        return TruthResult(matched=False, observed=observed)
    if not require_success:
        return TruthResult(matched=True, observed=observed)
    if conclusion == "success":
        return TruthResult(matched=True, observed=observed)
    raise RunFailedFastError(observed)


class RunFailedFastError(Exception):
    """``wait run --success`` hit a terminal-but-wrong conclusion.

    Carries the ``observed`` dict so the CLI layer can still emit it
    in the JSON envelope before exiting 4.
    """

    def __init__(self, observed: dict[str, Any]) -> None:
        super().__init__(f"run terminal conclusion: {observed.get('conclusion')}")
        self.observed = observed

File: src/wait.rs
use std::collections::BTreeMap;

use serde::Serialize;

/// Passing conclusions for `wait pr --state green`.
pub const PASSING_CONCLUSIONS: &[&str] = &["SUCCESS", "NEUTRAL", "SKIPPED"];
/// Still-waiting states for GitHub checks.
pub const STILL_WAITING_STATES: &[&str] = &["QUEUED", "IN_PROGRESS", "PENDING"];
/// Terminal run statuses for `wait run`.
pub const RUN_TERMINAL_STATUSES: &[&str] = &["completed"];

/// Error raised when a scope is unsupported, such as rulesets or merge queue.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct UnsupportedScopeError(pub String);

impl std::fmt::Display for UnsupportedScopeError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl std::error::Error for UnsupportedScopeError {}

/// Error raised when the input is missing or malformed.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct InvalidInputError(pub String);

impl std::fmt::Display for InvalidInputError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl std::error::Error for InvalidInputError {}

/// Error raised when `wait run --success` sees a terminal failure.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RunFailedFastError {
    /// Observed run snapshot.
    pub observed: BTreeMap<String, serde_json::Value>,
}

impl std::fmt::Display for RunFailedFastError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "run terminal conclusion: {}",
            self.observed
                .get("conclusion")
                .and_then(serde_json::Value::as_str)
                .unwrap_or("<unknown>")
        )
    }
}

impl std::error::Error for RunFailedFastError {}

/// Truth-condition evaluation result for `shipyard wait`.
#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct TruthResult {
    /// Whether the condition matched.
    pub matched: bool,
    /// JSON-serializable observed state.
    pub observed: BTreeMap<String, serde_json::Value>,
}

/// Evaluate whether a release satisfies the manifest condition.
pub fn evaluate_release(
    snapshot: Option<&serde_json::Value>,
    manifest: Option<&[String]>,
) -> Result<TruthResult, InvalidInputError> {
    let Some(snapshot) = snapshot else {
        return Ok(TruthResult {
            matched: false,
            observed: BTreeMap::from([("exists".to_owned(), serde_json::Value::Bool(false))]),
        });
    };

    let draft = snapshot
        .get("draft")
        .and_then(serde_json::Value::as_bool)
        .unwrap_or(false);
    if draft {
        return Ok(TruthResult {
            matched: false,
            observed: BTreeMap::from([
                ("exists".to_owned(), serde_json::Value::Bool(true)),
                ("draft".to_owned(), serde_json::Value::Bool(true)),
            ]),
        });
    }

    let assets = snapshot
        .get("assets")
        .and_then(serde_json::Value::as_array)
        .cloned()
        .unwrap_or_default();
    let observed_assets = assets
        .iter()
        .filter_map(|asset| asset.as_object())
        .map(|asset| {
            serde_json::json!({
                "name": asset.get("name").and_then(serde_json::Value::as_str).unwrap_or(""),
                "state": asset.get("state").and_then(serde_json::Value::as_str).unwrap_or(""),
                "size": asset.get("size").and_then(serde_json::Value::as_i64).unwrap_or(0),
            })
        })
        .collect::<Vec<_>>();

    let mut observed = BTreeMap::from([
        ("exists".to_owned(), serde_json::Value::Bool(true)),
        ("draft".to_owned(), serde_json::Value::Bool(false)),
        (
            "tag_name".to_owned(),
            snapshot
                .get("tag_name")
                .cloned()
                .unwrap_or(serde_json::Value::String(String::new())),
        ),
        (
            "assets".to_owned(),
            serde_json::Value::Array(observed_assets.clone()),
        ),
    ]);

    if let Some(manifest) = manifest.filter(|items| !items.is_empty()) {
        let by_name = observed_assets
            .iter()
            .filter_map(|asset| {
                let name = asset.get("name")?.as_str()?.to_owned();
                Some((name, asset))
            })
            .collect::<BTreeMap<_, _>>();

        let mut missing = Vec::new();
        let mut not_uploaded = Vec::new();
        for name in manifest {
            match by_name.get(name) {
                None => missing.push(serde_json::Value::String(name.clone())),
                Some(asset)
                    if asset.get("state").and_then(serde_json::Value::as_str)
                        != Some("uploaded") =>
                {
                    not_uploaded.push(serde_json::Value::String(name.clone()));
                }
                Some(_) => {}
            }
        }
        observed.insert(
            "manifest".to_owned(),
            serde_json::Value::Array(
                manifest
                    .iter()
                    .cloned()
                    .map(serde_json::Value::String)
                    .collect(),
            ),
        );
        observed.insert(
            "missing".to_owned(),
            serde_json::Value::Array(missing.clone()),
        );
        observed.insert(
            "not_uploaded".to_owned(),
            serde_json::Value::Array(not_uploaded.clone()),
        );
        return Ok(TruthResult {
            matched: missing.is_empty() && not_uploaded.is_empty(),
            observed,
        });
    }

    let matched = observed_assets
        .iter()
        .any(|asset| asset.get("state").and_then(serde_json::Value::as_str) == Some("uploaded"));
    Ok(TruthResult { matched, observed })
}

/// Evaluate whether a PR is green under classic required-check semantics.
pub fn evaluate_pr_green(
    snapshot: Option<&serde_json::Value>,
) -> Result<TruthResult, Box<dyn std::error::Error>> {
    let snapshot = snapshot.ok_or_else(|| InvalidInputError("PR not found".to_owned()))?;
    let merge_state = snapshot
        .get("mergeStateStatus")
        .and_then(serde_json::Value::as_str)
        .unwrap_or("")
        .to_ascii_uppercase();
    if merge_state.contains("RULESET") || merge_state == "MERGE_QUEUED" {
        return Err(Box::new(UnsupportedScopeError(
            "Rulesets / merge-queue governance isn't supported by `shipyard wait pr --state green` yet — see governance/profiles.py.".to_owned(),
        )));
    }

    let rollup = snapshot
        .get("statusCheckRollup")
        .and_then(serde_json::Value::as_array)
        .cloned()
        .unwrap_or_default();

    let mut checks = evaluate_pr_check_rollup(rollup);

    if checks.required_entries.is_empty() {
        checks.all_required_pass = snapshot
            .get("mergeable")
            .and_then(serde_json::Value::as_str)
            .unwrap_or("")
            .eq_ignore_ascii_case("MERGEABLE");
    }

    Ok(TruthResult {
        matched: checks.all_required_pass && !checks.any_still_waiting,
        observed: pr_green_observed(snapshot, merge_state, checks),
    })
}

struct PrGreenChecks {
    required_entries: Vec<serde_json::Value>,
    advisory_entries: Vec<serde_json::Value>,
    all_required_pass: bool,
    any_still_waiting: bool,
}

fn evaluate_pr_check_rollup(rollup: Vec<serde_json::Value>) -> PrGreenChecks {
    let mut checks = PrGreenChecks {
        required_entries: Vec::new(),
        advisory_entries: Vec::new(),
        all_required_pass: true,
        any_still_waiting: false,
    };

    for entry in rollup {
        let Some(entry) = entry.as_object() else {
            continue;
        };
        let observed = observed_pr_check(entry);
        if !observed.required {
            checks.advisory_entries.push(observed.value);
            continue;
        }
        checks.required_entries.push(observed.value);
        if observed.waiting {
            checks.any_still_waiting = true;
            checks.all_required_pass = false;
        } else if !observed.passing {
            checks.all_required_pass = false;
        }
    }
    checks
}

struct ObservedPrCheck {
    value: serde_json::Value,
    required: bool,
    waiting: bool,
    passing: bool,
}

fn observed_pr_check(entry: &serde_json::Map<String, serde_json::Value>) -> ObservedPrCheck {
    let name = entry
        .get("name")
        .or_else(|| entry.get("context"))
        .and_then(serde_json::Value::as_str)
        .unwrap_or("")
        .to_owned();
    let state = upper_entry_value(entry, "state");
    let conclusion = upper_entry_value(entry, "conclusion");
    let required = entry
        .get("isRequired")
        .and_then(serde_json::Value::as_bool)
        .unwrap_or(false);
    let waiting = STILL_WAITING_STATES.contains(&state.as_str())
        && !PASSING_CONCLUSIONS.contains(&conclusion.as_str());
    let passing = PASSING_CONCLUSIONS.contains(&conclusion.as_str());

    ObservedPrCheck {
        value: serde_json::json!({
            "name": name,
            "state": nullable_uppercase(state),
            "conclusion": nullable_uppercase(conclusion),
            "required": required,
        }),
        required,
        waiting,
        passing,
    }
}

fn upper_entry_value(entry: &serde_json::Map<String, serde_json::Value>, key: &str) -> String {
    entry
        .get(key)
        .and_then(serde_json::Value::as_str)
        .unwrap_or("")
        .to_ascii_uppercase()
}

fn nullable_uppercase(value: String) -> serde_json::Value {
    if value.is_empty() {
        serde_json::Value::Null
    } else {
        serde_json::Value::String(value)
    }
}

fn pr_green_observed(
    snapshot: &serde_json::Value,
    merge_state: String,
    checks: PrGreenChecks,
) -> BTreeMap<String, serde_json::Value> {
    BTreeMap::from([
        (
            "pr".to_owned(),
            snapshot
                .get("number")
                .cloned()
                .unwrap_or(serde_json::Value::Null),
        ),
        (
            "head_sha".to_owned(),
            snapshot
                .get("headRefOid")
                .cloned()
                .unwrap_or(serde_json::Value::String(String::new())),
        ),
        (
            "merge_state_status".to_owned(),
            nullable_uppercase(merge_state),
        ),
        (
            "checks".to_owned(),
            serde_json::Value::Array(checks.required_entries),
        ),
        (
            "advisory".to_owned(),
            serde_json::Value::Array(checks.advisory_entries),
        ),
    ])
}

/// Evaluate whether a PR reached `merged` or `closed`.
pub fn evaluate_pr_state(
    snapshot: Option<&serde_json::Value>,
    target_state: &str,
) -> Result<TruthResult, InvalidInputError> {
    let snapshot = snapshot.ok_or_else(|| InvalidInputError("PR not found".to_owned()))?;
    let state = snapshot
        .get("state")
        .and_then(serde_json::Value::as_str)
        .unwrap_or("")
        .to_ascii_uppercase();
    let merged = snapshot
        .get("merged")
        .and_then(serde_json::Value::as_bool)
        .unwrap_or(false);
    let matched = match target_state {
        "merged" => merged,
        "closed" => state == "CLOSED" || state == "MERGED",
        other => return Err(InvalidInputError(format!("unknown target state {other:?}"))),
    };
    Ok(TruthResult {
        matched,
        observed: BTreeMap::from([
            (
                "pr".to_owned(),
                snapshot
                    .get("number")
                    .cloned()
                    .unwrap_or(serde_json::Value::Null),
            ),
            ("state".to_owned(), serde_json::Value::String(state)),
            ("merged".to_owned(), serde_json::Value::Bool(merged)),
        ]),
    })
}

/// Evaluate whether a workflow run reached a terminal state.
pub fn evaluate_run(
    snapshot: Option<&serde_json::Value>,
    require_success: bool,
) -> Result<TruthResult, Box<dyn std::error::Error>> {
    let snapshot = snapshot.ok_or_else(|| InvalidInputError("run not found".to_owned()))?;
    let status = snapshot
        .get("status")
        .and_then(serde_json::Value::as_str)
        .unwrap_or("")
        .to_ascii_lowercase();
    let conclusion = snapshot
        .get("conclusion")
        .and_then(serde_json::Value::as_str)
        .unwrap_or("")
        .to_ascii_lowercase();
    let observed = BTreeMap::from([
        (
            "run_id".to_owned(),
            snapshot
                .get("databaseId")
                .or_else(|| snapshot.get("id"))
                .cloned()
                .unwrap_or(serde_json::Value::Null),
        ),
        (
            "status".to_owned(),
            serde_json::Value::String(status.clone()),
        ),
        (
            "conclusion".to_owned(),
            if conclusion.is_empty() {
                serde_json::Value::Null
            } else {
                serde_json::Value::String(conclusion.clone())
            },
        ),
    ]);

    if !RUN_TERMINAL_STATUSES.contains(&status.as_str()) {
        return Ok(TruthResult {
            matched: false,
            observed,
        });
    }
    if !require_success {
        return Ok(TruthResult {
            matched: true,
            observed,
        });
    }
    if conclusion == "success" {
        return Ok(TruthResult {
            matched: true,
            observed,
        });
    }

    Err(Box::new(RunFailedFastError { observed }))
}

#[cfg(test)]
mod tests {
    use super::{
        InvalidInputError, RunFailedFastError, UnsupportedScopeError, evaluate_pr_green,
        evaluate_pr_state, evaluate_release, evaluate_run,
    };

    fn rollup_entry(
        name: &str,
        conclusion: &str,
        state: &str,
        required: bool,
    ) -> serde_json::Value {
        serde_json::json!({
            "name": name,
            "conclusion": conclusion,
            "state": state,
            "isRequired": required,
        })
    }

    #[test]
    fn release_missing_snapshot_is_not_matched() {
        let result = evaluate_release(None, None).expect("release");
        assert!(!result.matched);
        assert_eq!(result.observed["exists"], serde_json::Value::Bool(false));
    }

    #[test]
    fn release_manifest_requires_all_assets_uploaded() {
        let snapshot = serde_json::json!({
            "tag_name": "v1",
            "draft": false,
            "assets": [
                {"name": "linux", "state": "uploaded", "size": 10},
                {"name": "darwin", "state": "starter", "size": 0}
            ]
        });
        let pending = evaluate_release(
            Some(&snapshot),
            Some(&["linux".to_owned(), "darwin".to_owned()]),
        )
        .expect("release");
        assert!(!pending.matched);
        assert_eq!(
            pending.observed["not_uploaded"],
            serde_json::json!(["darwin"])
        );
    }

    #[test]
    fn pr_green_matches_when_all_required_pass() {
        let snapshot = serde_json::json!({
            "number": 151,
            "headRefOid": "abc123",
            "state": "OPEN",
            "mergeable": "MERGEABLE",
            "mergeStateStatus": "CLEAN",
            "statusCheckRollup": [
                rollup_entry("Linux", "SUCCESS", "COMPLETED", true),
                rollup_entry("Windows", "SUCCESS", "COMPLETED", true),
                rollup_entry("macOS", "SUCCESS", "COMPLETED", true)
            ]
        });
        let result = evaluate_pr_green(Some(&snapshot)).expect("green");
        assert!(result.matched);
        assert_eq!(result.observed["head_sha"], "abc123");
        assert_eq!(
            result.observed["checks"].as_array().expect("checks").len(),
            3
        );
    }

    #[test]
    fn pr_green_ignores_advisory_failures() {
        let snapshot = serde_json::json!({
            "number": 1,
            "headRefOid": "x",
            "mergeable": "MERGEABLE",
            "mergeStateStatus": "CLEAN",
            "statusCheckRollup": [
                rollup_entry("Required", "SUCCESS", "COMPLETED", true),
                rollup_entry("Coverage", "FAILURE", "COMPLETED", false)
            ]
        });
        let result = evaluate_pr_green(Some(&snapshot)).expect("green");
        assert!(result.matched);
        assert_eq!(
            result.observed["checks"].as_array().expect("checks").len(),
            1
        );
        assert_eq!(
            result.observed["advisory"]
                .as_array()
                .expect("advisory")
                .len(),
            1
        );
    }

    #[test]
    fn pr_green_rulesets_raise_unsupported_scope() {
        let snapshot = serde_json::json!({
            "number": 1,
            "headRefOid": "x",
            "mergeable": "BLOCKED",
            "mergeStateStatus": "BLOCKED_BY_RULESET",
            "statusCheckRollup": []
        });
        let error = evaluate_pr_green(Some(&snapshot)).expect_err("unsupported");
        assert!(error.downcast_ref::<UnsupportedScopeError>().is_some());
    }

    #[test]
    fn pr_state_handles_merged_and_closed() {
        let merged = serde_json::json!({"number": 1, "state": "CLOSED", "merged": true});
        assert!(
            evaluate_pr_state(Some(&merged), "merged")
                .expect("merged")
                .matched
        );
        assert!(
            evaluate_pr_state(Some(&merged), "closed")
                .expect("closed")
                .matched
        );
    }

    #[test]
    fn run_success_fails_fast_on_terminal_wrong_conclusion() {
        let snapshot = serde_json::json!({
            "databaseId": 1,
            "status": "completed",
            "conclusion": "failure"
        });
        let error = evaluate_run(Some(&snapshot), true).expect_err("run failure");
        let error = error
            .downcast::<RunFailedFastError>()
            .expect("run failed fast error");
        assert_eq!(
            error
                .observed
                .get("conclusion")
                .and_then(serde_json::Value::as_str),
            Some("failure")
        );
    }

    #[test]
    fn run_non_terminal_is_not_matched() {
        let snapshot = serde_json::json!({
            "databaseId": 1,
            "status": "in_progress",
            "conclusion": null
        });
        let result = evaluate_run(Some(&snapshot), true).expect("run");
        assert!(!result.matched);
    }

    #[test]
    fn missing_pr_snapshot_raises_invalid_input() {
        let error = evaluate_pr_state(None, "merged").expect_err("missing pr");
        assert_eq!(error, InvalidInputError("PR not found".to_owned()));
    }
}
